src/adaptations/device-layer/include/Weave/DeviceLayer/internal/FactoryProvisioning.ipp

```diff
@@ -0,0 +1,249 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Contains non-inline method definitions for the FactoryProvisioningBase
+ *          template.
+ */
+
+#ifndef FACTORY_PROVISIONING_IPP
+#define FACTORY_PROVISIONING_IPP
+
+#include <Weave/DeviceLayer/internal/WeaveDeviceLayerInternal.h>
+#include <Weave/DeviceLayer/internal/FactoryProvisioning.h>
+
+namespace nl {
+namespace Weave {
+namespace DeviceLayer {
+namespace Internal {
+
+// Fully instantiate the default FactoryProvisioningBase class.
+template class FactoryProvisioningBase<FactoryProvisioning>;
+
+template<class DerivedClass>
+WEAVE_ERROR FactoryProvisioningBase<DerivedClass>::ProvisionDeviceFromRAM(uint8_t * memRangeStart, uint8_t * memRangeEnd)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+    uint8_t * dataStart;
+    size_t dataLen;
+
+    // Search the given RAM region for device provisioning data.  If found...
+    if (Derived()->LocateProvisioningData(memRangeStart, memRangeEnd, dataStart, dataLen))
+    {
+        // Wipe the provisioning data marker so that it will not be found again should the device reboot.
+        memset(dataStart - FactoryProvisioningData::kMarkerLen, 0, FactoryProvisioningData::kMarkerLen);
+
+        // Parse the provisioning data TLV and write the values to persistent storage.
+        TLV::TLVReader reader;
+        reader.Init(dataStart, dataLen);
+        err = Derived()->StoreProvisioningData(reader);
+        SuccessOrExit(err);
+
+        // Wipe the provisioning data itself.
+        memset(dataStart, 0, dataLen);
+    }
+
+exit:
+    return err;
+}
+
+template<class DerivedClass>
+bool FactoryProvisioningBase<DerivedClass>::LocateProvisioningData(uint8_t * memRangeStart, uint8_t * memRangeEnd, uint8_t * & dataStart, size_t & dataLen)
+{
+    using HashAlgo = Platform::Security::SHA256;
+
+    constexpr size_t kLenFieldLen = sizeof(uint32_t);
+
+    WeaveLogProgress(DeviceLayer, "Searching for factory provisioning data (0x%08" PRIX32 " - 0x%08" PRIX32 ")",
+            memRangeStart, memRangeEnd);
+
+    for (uint8_t * p = memRangeStart; p < memRangeEnd; p++)
+    {
+        // Search for the provisioning data marker within the given memory range.
+        p = (uint8_t *)memchr(p, FactoryProvisioningData::kMarker[0], memRangeEnd - p);
+        if (p == NULL)
+            break;
+        if (strncmp((char *)p, FactoryProvisioningData::kMarker, FactoryProvisioningData::kMarkerLen) != 0)
+            continue;
+
+        // Read the provisioning data length located immediately after the marker.
+        dataLen = (size_t)Encoding::LittleEndian::Get32(p + FactoryProvisioningData::kMarkerLen);
```

Comment:
I couldn't seem to find an identical write of this value in implementation. I might have just missed it. Is there or should there be documentation that this must be serialized in little endian format?

Reply:
Serialization happens the the python code here: [WeaveFactoryProvisioningTool.py:L276](https://github.com/openweave/openweave-core/blob/995b7d8054f5690e670e7d96e4067f4a9622a60f/src/tools/factory-prov-tool/WeaveFactoryProvisioningTool.py#L276)

Reply:
Thanks.